astropy/utils/masked/core.py

```diff
@@ -902,7 +903,7 @@ def __array_function__(self, function, types, args, kwargs):
             except NotImplementedError:
                 return self._not_implemented_or_raise(function, types)
 
-            if not isinstance(dispatched_result, tuple):
+            if not isinstance(dispatched_result, FunctionHelperReturnTuple):
```

Comment:
I see why you wrote it like this, but my sense is that really my design was just wrong. A dispatched function should just return a result, not possibly some combination - that should more logically be a different type of dispatched function.

I'll need to have a look at how often each case is used...

Reply:
In any case I think we should prioritise https://github.com/astropy/astropy/pull/15925, but yes, my patch is clearly not working as intended anyway so this might be a good time to re-evaluate the design here :)

Reply:
@mhvk is it okay if I drop the second commit for now ? I think it should make it clearer what problems still need solving.

Reply:
Yes, makes sense - let's fix the easy things while we think about the slightly harder ones!

Reply:
here we go !